Source/Provers/SMTLib/ProverInterface.cs

```diff
@@ -347,15 +388,80 @@ private void PrepareCommon()
       {
         var axioms = ctx.Axioms;
         var nary = axioms as VCExprNAry;
+	List<VCExpr> axiomList = new List<VCExpr>();
         if (nary != null && nary.Op == VCExpressionGenerator.AndOp)
-          foreach (var expr in nary.UniformArguments)
-          {
-            var str = VCExpr2String(expr, -1);
-            if (str != "true")
-              AddAxiom(str);
+          foreach (var expr in nary.UniformArguments) {
+            axiomList.Add(expr);
           }
         else
-          AddAxiom(VCExpr2String(axioms, -1));
+          axiomList.Add(axioms);
+
+        // Separate function definitions from other axioms.  Function definitions must be processed
+        // first; otherwise, processing an axiom that uses the function definition can create a declaration
+        // resulting in the symbol being introduced twice (a definition and a declaration).
+        Dictionary<Function, VCExpr> functionDefinitionMap = new Dictionary<Function, VCExpr>();
+        Stack<Function> functionDefs = new Stack<Function>();
+        Stack<VCExpr> otherAxioms = new Stack<VCExpr>();
+        for (int i = axiomList.Count-1; i >= 0; i--) {
+          Function f = IsFunctionDef(axiomList[i]);
```

Comment:
Following the similar C# operators, I would expect `is` to return a boolean (`true` if it is, and `false` if it ain't) and `as` to return a reference (the original value if it is, and `null` if it ain't). Therefore, I suggest renaming this function to `AsFunctionDef`.

Reply:
Done